[BE-13] feat: 스탭 공고 지원 API 구현

Woojae-Jeong opened the pull request:
## 작업한 내용은 무엇인가요?
- 스탭 공고 지원 API를 구현하였습니다

## 어떻게 작업했나요?
### 1. 공고 내 질문 여부 확인, 질문 조회 기능 구현
요청으로 들어온 공고를 저장하기 앞서서, 공고에는 질문이 있을 수도 없을 수도 있습니다. 따라서 공고에 추가 질문이 없는 경우에는 그저 지원 내역만 저장하며, 추가 질문이 있는 경우에는 `Request`로 들어온 `질문id`와 `공고id`가 존재하는지 확인하여 저장을 하게 됩니다. `공고id`가 추가적으로 필요한 이유는, 혹여 지원하는 공고가 가지고 있는 `질문`과 `Request`에서 답변한 `질문`이 서로 맞지 않게 오는 잘못된 요청이 있을 수도 있다고 생각하였습니다. 

### 2. ApplicationRecord, QuestionAnswer 엔티티 구현
스탭 지원 내역을 의미하는 `ApplicationRecord`, 지원하며 유저가 질문에 답변한 내용을 `QuestionAnswer` 로 구현하였습니다. 

## 리뷰어가 중점적으로 봐야할부분은 무엇인가요?
위의 1번과 2번에 대해서 로직상 이상은 없는지, 엔티티는 잘 설계 했는지 점검 부탁드립니다

## 기타 사항

## 관련 이슈

closes #24

Woojae-Jeong committed:
feat: 질문 여부 확인 및 질문 조회 기능 구현

Woojae-Jeong committed:
feat: 스탭 공고 지원 기능 구현

Woojae-Jeong committed:
feat: 스탭 공고 지원 API 구현

Woojae-Jeong committed:
refactor: UserId 필드를 삭제하고 ApplicationId로 변경

지원서와 스탭 공고간의 관계가 있는 것이므로, userId가 아닌 applicationId를 추가

Woojae-Jeong committed:
refactor: 지원서 조회 기능 구현

aiaiaiai1 commented on src/main/java/guesthouse/application/domain/model/QuestionAnswer.java:
```diff
@@ -0,0 +1,31 @@
+@Getter
+@Entity
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+public class QuestionAnswer {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @Column(nullable = false)
+    private Long applicationRecordId;
+
+    @Column(nullable = false)
+    private Long questionId;
+
+    @Column(nullable = false)
+    private String content;
+
+    public QuestionAnswer(Long applicationRecordId, Long questionId, String content) {
+        this.applicationRecordId = applicationRecordId;
```
다시 보니까 question을 id가 아닌 question String 으로 저장해야할거 같습니다. (+ index?)
지원서가 제출되고나서 사장님이 공고의 질문을 바꿔버리면 제출된 지원서의 질문도 바뀌어버리게 될것 같아요

Woojae-Jeong replied:
넵! `question`은 `String` 형태로 저장되게끔 리팩토링 하였습니다!

aiaiaiai1 commented on src/main/java/guesthouse/application/domain/model/ApplicationRecord.java:
```diff
@@ -17,10 +17,10 @@ public class ApplicationRecord {
+    public ApplicationRecord(Long recruitmentId, Long applicationId) {
         this.staffRecruitmentId = recruitmentId;
-        this.userId = userId;
+        this.applicationId = applicationId;
```
여기서도 동일하게 지원서의 id참조가 아닌 
제출한 지원서의 스냅샷을 저장해야할거 같습니다

Woojae-Jeong replied:
넵! 지원서의 스냅샷이 저장되게끔 리팩토링 하였습니다!
다만 지원서의 모든 내용이 담기는 것이 아닌 꼭 필요한 내용들만 스냅샷 형태로 저장되게끔 하였습니다

aiaiaiai1 reviewed (changes requested):
사용자 지원서를 제출하고나서 자신의 지원서를 수정하는 일과
사장님이 공고의 질문을 올리고난후 다시 공고의 질문을 수정하는 일이있을것 같아요
관련해서 커멘트 남겼습니다!

Woojae-Jeong committed:
refactor: QuestionAnswer의 질문은 스냅샷 형태로 수정

지원자가 공고글을 보고 지원한 후, 사장님이 질문글을 수정할 경우 올바르게 답변이 매칭이 안될 수 있다. 따라서, 질문에 대한 답변을 저장할 때 question 내용을 연관관계를 통한 참조 방식이 아닌 내용을 저장하는 즉, 스냅샷 저장 방식으로 리팩토링 하였습니다.

Woojae-Jeong committed:
refactor: ApplicationRecord의 지원서 내용은 스냅샷 형태로 수정

지원자가 공고글을 보고 지원한 후, 공통 지원서를 수정하게 될 수 있다. 수정을 하더라도, 지원당시의 지원서 내용이 보이게끔 스냅샷이 저장되도록 리팩토링을 진행하였습니다

aiaiaiai1 reviewed (approved):
확인했습니다~!

Woojae-Jeong committed:
Merge branch 'dev' into feat-#13
